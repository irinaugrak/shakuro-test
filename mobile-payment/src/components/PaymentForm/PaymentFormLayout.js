import React, {PureComponent} from 'react';
import {Field, reduxForm} from 'redux-form';
import ReactSVG from 'react-svg';
import {CSSTransition} from 'react-transition-group';
import PropTypes from 'prop-types';

import InputElement from '../FormElements/InputElement';
import InputPhoneElement from '../FormElements/InputPhoneElement';
import SuccessMessage from '../FormElements/SuccessMessage';
import ErrorMessage from '../FormElements/ErrorMessage';
import {phoneStrClear} from '../../utils/formFunctions';

import form_back_img from '../../assets/images/form-back.svg';

export const amountNormalize = (min, max) => (value, previousValue) => {

  if (!(/^\d+$/).test(value) && value !== '') {
    return previousValue;
  }

  if (parseInt(value, 10) < min || parseInt(value, 10) > max) {
    return previousValue;
  }

  return value;
};

const validate = values => {
  const errors = {};
  const errorsMessages = {
    required: 'Обязательное поле',
    badPhoneFormat: 'Неверный формат номера телефона'
  };

  if (!values.phone) {
    errors.phone = errorsMessages.required;

  } else if (phoneStrClear(values.phone).length < 11) {
    errors.phone = errorsMessages.badPhoneFormat;
  }

  if (!values.amount) {
    errors.amount = errorsMessages.required;
  }

  return errors;
};

class PaymentFormLayout extends PureComponent {

  componentDidMount() {
    this.props.initialize({operator: this.props.selectedOperator});
  }

  render() {
    const {
        handleSubmit,
        onPaymentFormBackClick,
        isDataLoading,
        submittedData,
        isDataSubmitError,
        onPaymentFormErrorMessageCloseClick
      } = this.props;
    const mobilePaymentPayWrapper =
      isDataLoading ?
      <div className='mobile-payment__pay-wrapper mobile-payment__pay-wrapper--loading'>
        <div className='spinner'/>
      </div>
      :
      submittedData ?
        <div className='mobile-payment__pay-wrapper'>
          <SuccessMessage submittedData={submittedData}/>
        </div>
        :
        isDataSubmitError ?
          <div className='mobile-payment__pay-wrapper'>
            <ErrorMessage onFormErrorMessageCloseClick={onPaymentFormErrorMessageCloseClick}/>
          </div>
          :
          <div className='mobile-payment__pay-wrapper'>
            <ReactSVG
              path={form_back_img}
              className='form__back'
              style={{
                width: 24,
                height: 16
              }}
              onClick={onPaymentFormBackClick}
            />
            <form
              className='form'
              onSubmit={handleSubmit}
            >
              <Field
                component={InputElement}
                type='text'
                name='operator'
                label='Оператор'
                format={value => value ? value.label : ''}
                disabled
              />
              <Field
                component={InputPhoneElement}
                type='text'
                name='phone'
                label='Номер телефона'
              />
              <Field
                component={InputElement}
                type='text'
                name='amount'
                label='Сумма платежа (до 1000 рублей)'
                normalize={amountNormalize(1, 1000)}
              />
              <button className='form__submit' type='submit'>Оплатить</button>
            </form>
          </div>;

    return (
      <div className='mobile-payment__pay'>
        <CSSTransition
          timeout={400}
          classNames='fade'
          in={submittedData !== null || isDataSubmitError !== false}
        >
          {mobilePaymentPayWrapper}
        </CSSTransition>
      </div>
    )
  }
}

export default reduxForm({
  form: 'paymentForm',
  validate
})(PaymentFormLayout);

PaymentFormLayout.propTypes = {
  handleSubmit: PropTypes.func.isRequired,
  onPaymentFormBackClick: PropTypes.func.isRequired,
  isDataLoading: PropTypes.bool.isRequired,
  submittedData: PropTypes.object,
  isDataSubmitError: PropTypes.bool.isRequired,
  onPaymentFormErrorMessageCloseClick: PropTypes.func.isRequired
};