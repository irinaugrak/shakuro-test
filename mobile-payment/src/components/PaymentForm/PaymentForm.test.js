import React from 'react';
import {reduxForm} from 'redux-form';
import renderer from 'react-test-renderer';
import {createStore} from 'redux';
import {Provider} from 'react-redux';
import { MemoryRouter } from 'react-router-dom';

import PaymentFormLayout from './PaymentFormLayout';
import {amountNormalize} from "./PaymentFormLayout";

const func = jest.fn();

const setPaymentFormLayoutParams = params => {

  return {
    isDataLoading: params.isDataLoading,
    isDataSubmitError: params.isDataSubmitError,
    submittedData: params.submittedData,
    onPaymentFormErrorMessageCloseClick: func,
    onPaymentFormBackClick: func
  }
};

const PaymentFormLayoutTestCaseCreate = testCase => {
  const store = createStore(() => ({}));
  const DecoratedPaymentFormLayout = reduxForm({form: 'paymentForm'})(PaymentFormLayout);
  const params = setPaymentFormLayoutParams(testCase);
  const element = renderer
    .create(
      <Provider store={store}>
        <MemoryRouter>
          <DecoratedPaymentFormLayout
            {...params}
          />
        </MemoryRouter>
      </Provider>
    )
    .toJSON();

  expect(element).toMatchSnapshot();
};

it('renders PaymentFormLayout correctly', () => {
  const testCases = [
    {
      isDataLoading: true,
      submittedData: null,
      isDataSubmitError: false
    },
    {
      isDataLoading: false,
      submittedData: null,
      isDataSubmitError: true
    },
    {
      isDataLoading: false,
      submittedData: {testData: 'testData'},
      isDataSubmitError: false
    },
    {
      isDataLoading: false,
      submittedData: null,
      isDataSubmitError: false
    }
  ];

  testCases.map(item => {
    PaymentFormLayoutTestCaseCreate(item);
  });
});

it('testing amount validation from PaymentFormLayout', () => {
  expect(amountNormalize(1, 1000)('10001', '1000')).toEqual('1000');
  expect(amountNormalize(1, 1000)('10000', '')).toEqual('');
  expect(amountNormalize(1, 1000)('10a', '10')).toEqual('10');
  expect(amountNormalize(1, 1000)('112', '11')).toEqual('112');
});


