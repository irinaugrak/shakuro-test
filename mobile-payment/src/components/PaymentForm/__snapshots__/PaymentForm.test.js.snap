// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders PaymentFormLayout correctly 1`] = `
<div
  className="mobile-payment__pay"
>
  <div
    className="mobile-payment__pay-wrapper mobile-payment__pay-wrapper--loading"
  >
    <div
      className="spinner"
    />
  </div>
</div>
`;

exports[`renders PaymentFormLayout correctly 2`] = `
<div
  className="mobile-payment__pay"
>
  <div
    className="mobile-payment__pay-wrapper"
  >
    <div
      className="form__error-message"
    >
      <p
        className="form__error-message-text"
      >
        Ваш платеж не прошел
      </p>
      <p
        className="form__error-message-one-more"
        onClick={[Function]}
      >
        Попробовать снова
      </p>
    </div>
  </div>
</div>
`;

exports[`renders PaymentFormLayout correctly 3`] = `
<div
  className="mobile-payment__pay"
>
  <div
    className="mobile-payment__pay-wrapper"
  >
    <div
      className="form__success-message"
    >
      <p
        className="form__success-message-text"
      >
        Успешно отправлены данные:
      </p>
      <p
        className="form__success-message-data"
      >
        {"testData":"testData"}
      </p>
      <a
        className="form__success-message-link"
        href="/"
        onClick={[Function]}
      >
        На главную страницу
      </a>
    </div>
  </div>
</div>
`;

exports[`renders PaymentFormLayout correctly 4`] = `
<div
  className="mobile-payment__pay"
>
  <div
    className="mobile-payment__pay-wrapper"
  >
    <div
      className="form__back"
      onClick={[Function]}
      style={
        Object {
          "height": 16,
          "width": 24,
        }
      }
    />
    <form
      className="form"
      onSubmit={[Function]}
    >
      <div
        className="form__field"
      >
        <label
          className="form__field-label"
        >
          Оператор
        </label>
        <div
          className="form__field-input-item"
        >
          <input
            className="form__field-input"
            disabled={true}
            name="operator"
            onBlur={[Function]}
            onChange={[Function]}
            onDragStart={[Function]}
            onDrop={[Function]}
            onFocus={[Function]}
            type="text"
            value=""
          />
        </div>
      </div>
      <div
        className="form__field"
      >
        <label
          className="form__field-label"
        >
          Номер телефона
        </label>
        <div
          className="form__field-input-item"
        >
          <input
            className="form__field-input"
            mask="+1 (111) 111-11-11"
            maxLength={18}
            name="phone"
            onBlur={[Function]}
            onChange={[Function]}
            onDragStart={[Function]}
            onDrop={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onKeyPress={[Function]}
            onPaste={[Function]}
            placeholder=""
            size={18}
            type="text"
            value=""
          />
        </div>
      </div>
      <div
        className="form__field"
      >
        <label
          className="form__field-label"
        >
          Сумма платежа (до 1000 рублей)
        </label>
        <div
          className="form__field-input-item"
        >
          <input
            className="form__field-input"
            disabled={false}
            name="amount"
            onBlur={[Function]}
            onChange={[Function]}
            onDragStart={[Function]}
            onDrop={[Function]}
            onFocus={[Function]}
            type="text"
            value=""
          />
        </div>
      </div>
      <button
        className="form__submit"
        type="submit"
      >
        Оплатить
      </button>
    </form>
  </div>
</div>
`;
